From: Sebastian Wick <sebastian.wick@redhat.com>
Date: Fri, 10 Apr 2026 15:44:44 +0200
Subject: tests: Check that flatpak-run fd-arguments do not leak to the
 command

flatpak-run takes a number of arguments which are file descriptor
numbers. Those file descriptors are supposed to set something up in the
way the instance gets spawned, but should never make it to the wrapper
command.

(cherry picked from commit 136452768368613f3712c2f7cba79a7c3e7bee96)

Origin: upstream, 1.16.6, commit:4491406f509707feb755982fdda24b2fd9234903
---
 tests/test-run-custom.sh | 25 +++++++++++++++++++++++--
 1 file changed, 23 insertions(+), 2 deletions(-)

diff --git a/tests/test-run-custom.sh b/tests/test-run-custom.sh
index 94c3241..4868b8b 100755
--- a/tests/test-run-custom.sh
+++ b/tests/test-run-custom.sh
@@ -24,7 +24,7 @@ set -euo pipefail
 skip_without_bwrap
 skip_revokefs_without_fuse
 
-echo "1..11"
+echo "1..12"
 
 # Use stable rather than master as the branch so we can test that the run
 # command automatically finds the branch correctly
@@ -176,4 +176,25 @@ exec 3< "${path}"
 ! run --ro-bind-fd=3 --command=bash org.test.Hello -c "echo baz > ${path}" > /dev/null
 exec 3>&-
 
-ok "bind-fd and ro-bind-fd"
\ No newline at end of file
+ok "bind-fd and ro-bind-fd"
+
+exec 3< custom-app/files
+exec 4< custom-runtime/files
+exec 5< "${path}"
+exec 6< "${path}"
+run --app-fd=3 --usr-fd=4 --bind-fd=5 --ro-bind-fd=6 \
+    --command=sh org.test.Hello \
+    -c 'for fd in $(ls /proc/self/fd); do readlink -f /proc/self/fd/$fd; done' > hello_out
+exec 6>&-
+exec 5>&-
+exec 4>&-
+exec 3>&-
+
+wd="$(readlink -f .)"
+while read fdpath; do
+  if [[ "$fdpath" == "$wd"* && "$fdpath" != "$wd/hello_out" ]]; then
+    assert_not_reached "A fd for '$fdpath' unexpectedly made it to the app"
+  fi
+done < hello_out
+
+ok "check no fd leak"
\ No newline at end of file
